File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length, n = matrix[0].length;
        int l = 0, r = m-1;
        Integer inRow = null;
        while(l<=r){
            int mid = (l+r)/2;
            if(target >= matrix[mid][0] && target <= matrix[mid][n-1]){
                inRow = mid;
                break;
            }
            else if(target < matrix[mid][0]){
                r = mid-1;
            }
            else{
                l = mid+1;
            }
        }
        
        if(inRow == null)
            return false;
        
        l = 0;
        r = n-1;
        while(l<=r){
            int mid = (l+r)/2;
            if(target == matrix[inRow][mid])
                return true;
            else if(target < matrix[inRow][mid])
                r = mid-1;
            else
                l = mid+1;
        }
        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = matrix[0].size();
        int row_l = -1, row_r = m;
        while(row_r - row_l > 1){
            int row_m = (row_r + row_l)/2;
            if(target < matrix[row_m][0])
                row_r = row_m;
            else if(target > matrix[row_m][n-1])
                row_l = row_m;
            else{
                auto found = *lower_bound(matrix[row_m].begin(), matrix[row_m].end(), target);
                if(found == target)
                    return true;
                else
                    return false;
            }
        }
        return false;
    }
};